Locks and queue for linking

geekbeast opened the pull request:
In Progress.

geekbeast committed:
Locks and queue for linking

geekbeast committed:
Implement linking queues

you-leee commented on src/main/kotlin/com/openlattice/linking/controllers/RealtimeLinkingController.kt:
```diff
@@ -29,6 +25,9 @@ class RealtimeLinkingController(
     private lateinit var authz: AuthorizationManager
 
+    @Inject
+    private lateinit var backgroundLinkingService: BackgroundLinkingService
```
I don't think the RealtimeLinkingController needs the BackgroundLinkingService

geekbeast replied:
I am adding some APIs for controlling linking.

makosblade commented on src/main/kotlin/com/openlattice/linking/BackgroundLinkingService.kt:
```diff
@@ -0,0 +1,347 @@
+internal const val MINIMUM_SCORE = 0.75
+
+/**
+ * Performs realtime linking of individuals as they are integrated ino the system.
```
🤔

makosblade reviewed (approved):
LGTM

geekbeast committed:
Streamline distributed locking.

We now lock at the level of the object being linked. This makes it easy to horizontally scale linking as required by allowing disjoint clusters to be processed independently.
The key observation is that main decision is merge or create. You don't want to merge into a cluster that is potentially being created or updated so if you block to a locked element
we simply skip processing that element. Still need to handle the merge case where two candidates that don't block to each other, but have a cluster in common, may attempt to merge into
the same cluster.

geekbeast committed:
Lock cluster for updates in postgres

geekbeast commented:
This ready to go, implementation of linking management apis will be in a future PR.
